storage/src/adb/any/sync/client.rs

```diff
@@ -0,0 +1,705 @@
+            },
+            config.lower_bound_ops,
+        )
+        .await
+        .map_err(adb::Error::JournalError)
+        .map_err(Error::Adb)?;
+
+        Ok(Client::FetchData {
+            metrics: Metrics::new(config.context.clone()),
+            config,
+            log,
+            pinned_nodes: None,
+        })
+    }
+
+    /// Process the next step in the sync process
+    async fn step(self) -> Result<Self, Error> {
+        match self {
+            Client::FetchData {
+                mut config,
+                log,
+                mut pinned_nodes,
+                metrics,
+            } => {
+                // Get current position in the log
+                let log_size = log.size().await.unwrap();
+
+                // Calculate remaining operations to sync (inclusive upper bound)
+                let remaining_ops = if log_size <= config.upper_bound_ops {
+                    config.upper_bound_ops - log_size + 1
+                } else {
+                    // We're at/past the target.
+                    warn!(
+                        log_size,
+                        upper_bound = config.upper_bound_ops,
+                        "Sync target exceeded"
+                    );
+                    return Err(Error::InvalidState);
+                };
+
+                let batch_size = std::cmp::min(config.fetch_batch_size.get(), remaining_ops);
+                let batch_size = NonZeroU64::new(batch_size).ok_or(Error::InvalidState)?;
+
+                debug!(
+                    target_hash = ?config.target_hash,
+                    lower_bound_pos = config.lower_bound_ops,
+                    upper_bound_pos = config.upper_bound_ops,
+                    current_pos = log_size,
+                    remaining_ops = remaining_ops,
+                    batch_size = batch_size.get(),
+                    "Fetching proof and operations"
+                );
+
+                // Get proof and operations from resolver with timing
+                let GetProofResult {
+                    proof,
+                    operations,
+                    success_tx,
+                } = {
+                    let _timer = metrics.fetch_duration.timer();
+                    config.resolver.get_proof(log_size, batch_size).await?
+                };
+
+                let operations_len = operations.len() as u64;
+
+                // Validate that we didn't get more operations than requested
+                // or that we didn't get an empty proof. We should never get an empty proof
+                // because we will never request an empty proof (i.e. a proof over an empty database).
+                if operations_len > batch_size.get() || operations_len == 0 {
+                    metrics.invalid_batches_received.inc();
+                    let _ = success_tx.send(false);
+                    if metrics.invalid_batches_received.get() > config.max_retries {
```

Comment:
In other areas of the code we never give up, I think we should remove this feature.

If we do keep it, I think it should be over failures of a particular batch (not over all batches).